Prose/ProseLib/Sources/AuthenticationFeature/MFA6DigitsView.swift

```diff
@@ -0,0 +1,275 @@
+//
+//  MFA6DigitsView.swift
+//  Prose
+//
+//  Created by Rémi Bardon on 13/06/2022.
+//
+
+import AppLocalization
+import ComposableArchitecture
+import SwiftUI
+import SwiftUINavigation
+import TcaHelpers
+
+private let l10n = L10n.Authentication.Mfa.self
+
+// MARK: - View
+
+public struct MFA6DigitsView: View {
+    public typealias State = MFA6DigitsState
+    public typealias Action = MFA6DigitsAction
+
+    let store: Store<State, Action>
+    private var actions: ViewStore<Void, Action> { ViewStore(self.store.stateless) }
+
+    @FocusState private var focusedField: State.Field?
+
+    public var body: some View {
+        WithViewStore(self.store) { viewStore in
+            VStack(spacing: 32) {
+                Self.header()
+                fields(viewStore: viewStore)
+                mainButton(viewStore: viewStore)
+                footer(viewStore: viewStore)
+            }
+            .multilineTextAlignment(.center)
+            .controlSize(.large)
+            .textFieldStyle(.roundedBorder)
+            .buttonBorderShape(.roundedRectangle)
+            .padding(.horizontal, 48)
+            .padding(.vertical, 24)
+            .synchronize(viewStore.binding(\.$focusedField), self.$focusedField)
+            .alert(self.store.scope(state: \.alert), dismiss: .alertDismissed)
+            .onAppear { actions.send(.onAppear) }
+            .onTapGesture { actions.send(.onGlobalTapGesture) }
+        }
+    }
+
+    static func header() -> some View {
+        VStack(spacing: 4) {
+            Text("👇")
+                .font(.system(size: 64))
+                .padding(.bottom, 20)
+            Text(l10n.Header.title)
+                .font(.system(size: 24, weight: .bold))
+            Text(l10n.Header.subtitle)
+                .foregroundColor(.secondary)
+                .font(.system(size: 16))
+        }
+        .accessibilityElement()
+        .accessibilityLabel(l10n.Header.subtitle)
+    }
+
+    func fields(viewStore: ViewStore<State, Action>) -> some View {
+        HStack(spacing: 4) {
+            ForEach(0..<6) { n in
+                digit(viewStore.state.digit(n), selected: viewStore.code.count == n)
+            }
+        }
+        // Hide fake fields in accessibility
+        .accessibilityHidden(true)
+        // Add a real `TextField` backing the interactions and accessibility
+        .background {
+            TextField(l10n.Form.OneTimeCode.placeholder, text: viewStore.binding(\.$code))
+                .focused($focusedField, equals: .textField)
+                .textContentType(.oneTimeCode)
+                .opacity(0)
```

Comment:
That was a cool idea with the hidden textfield.

There's just one thing that bothers me: You're filtering out non-numeric characters which is good. However if the user _does_ type a non-numeric character they land in a state that's non obvious. Suddenly they have to press backspace as many times as they've entered invalid characters, but don't see the result. It would be better if the contents of the hidden textfield would reflect what's on screen. 

Here's something that I've [found](https://stackoverflow.com/a/58736068) that does that. I don't know if there's an easier way in SwiftUI?

Reply:
Yeah, I don't know if there's an easy solution. I just told myself that users would not type letters so we'd be good for 99% of them.

I just realised I didn't set the keyboard layout to "decimal digits", maybe this can help (if not, at least it'll set the correct keyboard on iOS (if `.textContentType(.oneTimeCode)` doesn't do it already))

Reply:
We should either let the users type garbage in there or filter it out completely. No hidden states.

Reply:
Yeah, I agree

Reply:
Oh yeah, why did I filter at all?

Reply:
This seemed interesting, unfortunately, as it's an `Int`, it would not support having an empty field.

https://stackoverflow.com/a/70601917/10967642

Reply:
The ViewModifier looks quite nice though: https://stackoverflow.com/a/71822350

Reply:
Yesterday I gave a few tries at this, but unfortunately I can't override what the user inputs in the `TextField` 😕 The UI and the `Binding` are in fact two separate things, and even though I filter in the `Binding`, it doesn't filter in the field. I'm thinking about doing someting like filtering and then

```swift
DispatchQueue.main.async {
    realBinding.wrappedValue = filteredInput
}
```

so it overrides on next render.

Reply:
@nesium I managed to make it work, using https://stackoverflow.com/a/71093025/10967642.

https://user-images.githubusercontent.com/37386490/174240560-39e22324-0fd7-4e71-a610-ddd03130c7c6.mov

(The numbers don't react yet, I've just disabled them)

Reply:
I will also block to 6 characters

Reply:
It works 🥳 And everything is handled in the Reducer 🙂

https://user-images.githubusercontent.com/37386490/174247125-d4951cce-be48-413d-a701-4f26d141ed24.mov